File: Study Plans/30 Days of LC JavaScript Challenge/Day-6.js
// Problem :- 2626. Array Reduce Transformation
// Link :- https://leetcode.com/problems/array-reduce-transformation/?gio_link_id=nPN45jD9

// Solution:
/**
 * @param {number[]} nums
 * @param {Function} fn
 * @param {number} init
 * @return {number}
 */
var reduce = function(nums, fn, init) {
    if(nums.length === 0) return init;
    let val = fn(init, nums[0]), i = 1;
    while(i < nums.length) val = fn(val, nums[i++]);
    return val;
};

File: LeetCode/POTD/2025/January/Day-6.cpp
// Problem Link: https://leetcode.com/problems/minimum-number-of-operations-to-move-all-balls-to-each-box/

// Solution:
class Solution {
    int fillBalls(int& x, vector<int>& balls) {
        int ans = 0;
        for(auto &i : balls) ans += abs(x-i);
        return ans;
    }
public:
    vector<int> minOperations(string boxes) {
        int n = boxes.length();
        vector<int> ans(n, 0), id;
        for(int i = 0; i < n; i++) if(boxes[i] == '1') id.emplace_back(i);
        for(int i = 0; i < n; i++) ans[i] = fillBalls(i, id);
        return ans;
    }
};